core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java

```diff
@@ -100,6 +101,214 @@ RelDataType deriveCovarType(RelDataTypeFactory typeFactory,
   /** Whether the least restrictive type of a number of CHAR types of different
    * lengths should be a VARCHAR type. And similarly BINARY to VARBINARY. */
   boolean shouldConvertRaggedUnionTypesToVarying();
+
+  /**
+   * Infers the return type of a decimal addition. Decimal addition involves
+   * at least one decimal operand and requires both operands to have exact
+   * numeric types.
+   *
+   * @param type1 type of the first operand
+   * @param type2 type of the second operand
+   * @return the result type for a decimal addition.
+   */
+  default RelDataType deriveDecimalPlusType(RelDataTypeFactory typeFactory,
+                                            RelDataType type1, RelDataType type2) {
+    /**
```

Comment:
(style): it should be a regular comment, and not a javadoc one. Since this is the default behavior, might want to move it from comment to proper javadoc.